### ISSUE TITLE
Crash Report: Bad save data results in game crash immediately on startup

### ISSUE THREAD

**Issue by KindOfASpork** [FunkinCrew/Funkin#2743]
## Describe the bug
The game crashes immediately after launching

## To Reproduce
1. Launch the game
## Expected behavior
The game should run as normal

## Screenshots/Video
The game's output from the terminal (personal info scrubbed out)
![image](https://github.com/FunkinCrew/Funkin/assets/91402113/c1d72c94-e3b8-40cb-9710-5ce168fdeceb)

## Desktop
 - OS: Windows 10
 - Browser: n/a (this only affects the downloadable build)
 - Version: 0.4.0

## Additional context
- n/a

**Comment by EliteMasterEric** [comment]
I don't see any info that suggests a crash happened. Do you have a log file?

**Comment by KindOfASpork** [comment]
> I don't see any info that suggests a crash happened. Do you have a log file?

The game doesn't output a log file for this crash specifically, weirdly enough.
This is a video of the crash happening. Everything in the terminal screenshot earlier is all of what the game outputs before closing.

https://github.com/FunkinCrew/Funkin/assets/91402113/857441e5-eb56-4c12-9e95-9d011dfc7063

**Comment by ninjamuffin99** [comment]
hmm try current version of game and does it still occur?

**Comment by KindOfASpork** [comment]
> hmm try current version of game and does it still occur?

The game still crashes.
This is all taken from a fresh build of 0.4.1 straight from itch.

https://github.com/FunkinCrew/Funkin/assets/91402113/422e0d8c-f463-45a4-b7d5-251f1bd17a9b
![image](https://github.com/FunkinCrew/Funkin/assets/91402113/cd6d8751-18f1-48e0-b033-67d0315f1f4b)

**Comment by NotHyper-474** [comment]
According to this guy in Itch.io this gets solved if you delete your save data. Odd.

![Screenshot_20240613-093250_Chrome.jpg](https://github.com/FunkinCrew/Funkin/assets/40342021/2ddf9d30-2124-4c2f-820d-535ae91e14d2)

**Comment by KindOfASpork** [comment]
> According to this guy in Itch.io this gets solved if you delete your save data. Odd.
> 
> ![Screenshot_20240613-093250_Chrome.jpg](https://private-user-images.githubusercontent.com/40342021/339360233-2ddf9d30-2124-4c2f-820d-535ae91e14d2.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyODU3NzUsIm5iZiI6MTcxODI4NTQ3NSwicGF0aCI6Ii80MDM0MjAyMS8zMzkzNjAyMzMtMmRkZjlkMzAtMjEyNC00YzJmLTgyMGQtNTM1YWU5MWUxNGQyLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjEzVDEzMzExNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJjZjBlM2RhNWNkNTI1YTllMTE1NTc2ZmU5NGM2ZmE0Y2JlZmE3NDI2YzBmMmNhZDZjY2JiZGE2NWE1MGFhNDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.HRWKbWAOi-8aZnUaXmpQgQTgswxq3F54H0lRf3eoLPc)

I tried this workaround, and 0.4 onwards boots up properly now.
So I guess it might have something to do with 0.4 not reading older saves correctly?

**Comment by NotHyper-474** [comment]
> I tried this workaround, and 0.4 onwards boots up properly now.
> So I guess it might have something to do with 0.4 not reading older saves correctly?

Could be the case. Do you happen to still have the save file backed up? The devs could debug it and see what's wrong

**Comment by EliteMasterEric** [comment]
Correct, if you have your save data we can diagnose what's wrong with it

**Comment by TheWooomy** [comment]
> Correct, if you have your save data we can diagnose what's wrong with it

I'm having the same issue and here's my save data if it helps
[Funkin1.zip](https://github.com/user-attachments/files/15830113/Funkin1.zip)

**Comment by mikolka9144** [comment]
Sorry for coming so late and I'm so sorry for causing this mess.


## Affected version
This bug is caused by older versions of my DDTO HUD (versions before 1.0) and affects all versions of FNF from v0.4 

## Cause
It's caused by the game's inability to properly handle maps being present in a game's save (FlxG.save).

## How to reproduce
Simply create a script, that will put an empty String map to the game's save.
Example script:
```
import funkin.modding.module.Module;
import flixel.FlxG;
import haxe.ds.StringMap;

class Settings extends Module {
    var DEFAULT:StringMap<Dynamic> = new StringMap();

    public function new() {
        super("Doki Settings",30);
        if(FlxG.save.data.dokiProps == null || !Std.isOfType(FlxG.save.data.dokiProps, StringMap)) 
        {
            FlxG.save.data.dokiProps = DEFAULT;
            FlxG.save.flush();
        }
    }

}

```

And run the game **twice** (crash will happen on the second run)

## Fix
My fix was to implement a special case for cloning maps:

File: src/thx/Dynamics.hx in "thx.core"
```
public static function clone(v:Dynamic, ?cloneInstances = false):Dynamic {
		switch (Type.typeof(v)) {
			case TNull:
				return null;
			case TInt, TFloat, TBool, TEnum(_), TUnknown, TFunction:
				return v;
			case TObject:
				return Objects.copyTo(v, {});
			case TClass(c):
				var name = Type.getClassName(c);
				switch (name) {
					case "Array":
						return (v : Array<Dynamic>).map(function(v) return clone(v, cloneInstances));
					case "String", "Date":
						return v;
					case "haxe.ds.StringMap": { //* the wonderful world of the base game
						 var src = (v : StringMap<Dynamic>);
						 var copied = new StringMap();
						 for (key in src.keys())
							 copied.set(key, clone(src.get(key),cloneInstances));
						 return copied; // more generic implementation might be possible
						}
					default:
						if (cloneInstances) {
							var o = Type.createEmptyInstance(c);
							for (field in Type.getInstanceFields(c))
								Reflect.setField(o, field, clone(Reflect.field(v, field), cloneInstances));
							return o;
						} else {
							return v;
						}
				}
		}
	}

```
Although there might be a better way to fix this.

And I thought this only affected me..

**Comment by NotHyper-474** [comment]
The bad save file still causes a crash in v0.5.1

**Comment by Hundrec** [comment]
This is from an old version, so I'll mark it as `stale`.

**Comment by mikolka9144** [comment]
The issue still persists in 0.6.2
"""
"Invalid field:set
  in /usr/share/haxe/std/cpp/_std/Reflect.hx#41
  in thx/Dynamics.hx#184
  in thx/Objects.hx#185
  in thx/Objects.hx#156
  in funkin/save/migrator/SaveDataMigrator.hx#31
  in funkin/save/Save.hx#1020
  in funkin/save/Save.hx#58
  in Main.hx#106
  in Main.hx#75
"
"""
@Hundrec Could you please fix this issue for good?

**Comment by Hundrec** [comment]
Can you check if it still happens in 0.6.3?

**Comment by mikolka9144** [comment]
> Can you check if it still happens in 0.6.3?

I did, and the game crashed without creating any logs (only "logs" folder appeared)

**Comment by Hundrec** [comment]
If you can obtain a crash log generated by 0.6.3, please create a new issue and we'll try to figure out what's happening

**Comment by mikolka9144** [comment]
> If you can obtain a crash log generated by 0.6.3, please create a new issue and we'll try to figure out what's happening

The game literally crashed while trying to do that when I tested on 0.6.2

**Comment by NotHyper-474** [comment]
From the looks of it it seems more of an issue with the thx library than with Funkin.

**Comment by mikolka9144** [comment]
The fix has been merged now
